src/QMCWaveFunctions/Fermion/MultiSlaterDeterminantFast.cpp

```diff
@@ -816,9 +867,15 @@ void MultiSlaterDeterminantFast::evaluateDerivativesWF(ParticleSet& P,
           const RealType* restrict CSFexpansion_p = CSFexpansion->data();
           for (int k = 0; k < (*DetsPerCSF)[ip]; k++)
           {
-            size_t upC = (*C2node)[0][cnt];
-            size_t dnC = (*C2node)[1][cnt];
-            cdet += CSFexpansion_p[cnt] * detValues_up[upC] * detValues_dn[dnC] * psiinv;
+            ValueType t = CSFexpansion_p[cnt] * psiinv;
+            for (size_t id = 0; id < Dets.size(); id++)
+            {
+              size_t spinC = (*C2node)[id][cnt];
+              // assume that evaluateLog has been called in opt routine before
+              ValueVector_t& detValues_spin = Dets[id]->detValues;
+              t *= detValues_spin[spinC];
```

Comment:
better to avoid spinC and detValues_spin

Reply:
Done

Reply:
Not done?